src/AWS.jl

```
@@ -210,8 +216,12 @@ end
 # Needs to be included after the definition of struct otherwise it cannot find them
 include("AWSServices.jl")
 
-function generate_service_url(aws::AbstractAWSConfig, service::String, resource::String)
-    SERVICE_HOST = "amazonaws.com"
+function generate_service_url(aws::AbstractAWSConfig, service::String, resource::String; s3_acceleration::Bool=use_s3_acceleration[])
+    # S3 virtualhosts cannot have periods in their bucket names, default to path style in these cases
+    if service == "s3" && !contains(resource, ".")
```

What if `resource` contains a `.` and `s3_acceleration == true`? Should we throw an error? Or rely on the fact that amazon doesn't allow that already?

---

This comment just caught an issue here (I'll be adding tests if we agree on this solution). The `resource` passed in here would be something like,

> mattbr-open-world/path/to/some/file.jpg

In this case we want to parse out the bucket name, `mattbr-open-world`, and check that this does not contain a period. From the [documentation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html) virtualhost will not work if the bucket contains a period, however it is a valid character for bucket names.

In the case where it does contain a period, default back to the path style requests instead.

---

> From the [documentation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html) virtualhost will not work if the bucket contains a period, however it is a valid character for bucket names.

I think it does work with HTTP (but I don't think we support non-HTTPS connections) and if you don't perform the standard certificate validation. Better info here I think: https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html

I believe this is why SDKs support forcing `virtual` mode.